.github/workflows/actions/build-version/action.yml

```diff
@@ -0,0 +1,135 @@
+name: Build package
+inputs:
+  crate:
+    description: crates to build
+    required: true
+    type: string
+  features:
+    description: features to enable
+    required: true
+    type: string
+  no_default_features:
+    description: disable default features
+    required: true
+    type: string
+  target_arch:
+    description: Target to build against
+    required: true
+    type: string
+  version:
+    description: Version to release
+    required: true
+    type: string
+
+outputs:
+  artifact:
+    description: "artifact"
+    value: ${{ steps.output.outputs.artifact }}
+
+runs:
+  using: composite
+  steps:
+    - uses: taiki-e/install-action@v2
+      if: runner.os != 'Linux'
+      with:
+        tool: cargo-auditable
+
+    - name: Install LLVM
+      if: runner.os == 'Windows'
+      run: choco install llvm
+      shell: bash
+
+    - name: Install LLVM
+      if: runner.os == 'macOS'
+      run: |
+        set -euxo pipefail
+        brew install llvm
+        LLVM_PREFIX="$(brew --prefix llvm)"
+        echo "${LLVM_PREFIX}/bin" >> "$GITHUB_PATH"
+        echo "LLVM_CONFIG_PATH=${LLVM_PREFIX}/bin/llvm-config" >> "$GITHUB_ENV"
+        echo "DYLD_LIBRARY_PATH=${LLVM_PREFIX}/lib:${LD_LIBRARY_PATH:-}" >> "$GITHUB_ENV"
+        echo "LIBCLANG_PATH=$(llvm-config --libdir)" >> "$GITHUB_ENV"
+        echo "LLVM_DIR=$(llvm-config --cmakedir)" >> "$GITHUB_ENV"
+      shell: bash
+
+    - name: Decide cargo args, tempdir and cargo root
+      env:
+        FEATURES: ${{ inputs.features }}
+        NO_DEFAULT_FEATURES: ${{ inputs.no_default_features }}
+      run: |
+        set -exuo pipefail
+        features="${FEATURES:-}"
+        if [ -z "$features" ]; then
+          feature_flag=""
+        else
+          feature_flag="--features"
+        fi
+        no_default_features=""
+        if [ "${NO_DEFAULT_FEATURES:-}" = "true" ]; then
+          no_default_features='--no-default-features'
+        fi
+        echo "CARGO_ARGS=$no_default_features $feature_flag $features" >> "$GITHUB_ENV"
+        echo "TEMPDIR=$(mktemp -d)" >> "$GITHUB_ENV"
+        echo CARGO_HOME="$(mktemp -d 2>/dev/null || mktemp -d -t 'cargo-root')" >> "$GITHUB_ENV"
+      shell: bash
+
+    - name: Build packages
+      if: runner.os != 'Linux'
+      run: ./build-version.sh
+      shell: bash
+      env:
+        CRATE: ${{ inputs.crate }}
+        VERSION: ${{ inputs.version }}
+        TARGET_ARCH: ${{ inputs.target_arch }}
+
+    - name: Set up Docker Buildx
+      if: runner.os == 'Linux'
+      uses: docker/setup-buildx-action@v3
+
+    - name: Build docker image for package building
+      if: runner.os == 'Linux'
+      uses: docker/build-push-action@v5
+      with:
+        context: .
+        load: true
+        tags: package-builder:latest
+        cache-from: type=gha
```

Comment:
it's a bit scary that we can write to a cache here and potentially have it read by another package's build

Reply:
This is for building `Dockerfile` written by me, it only downloads the dependencies and never do any `cargo build`

Reply:
The artifact storage seems to be sandboxed per run or something, and actions/cache is per branch, with the ability to read from your base branch's cache but not write to it (unless you set something in your permissions, here). In our case I think all builds on the same arch happen from the same branch, so I guess we should treat anything we read from the cache as untrusted? 

I don't think it matters who wrote the code that is writing to the cache: if the cache is writeable from an action that's happening on this worker (for this branch) then we can't trust anything that's returned from the cache.

Reply:
> I don't think it matters who wrote the code that is writing to the cache: if the cache is writeable from an action that's happening on this worker (for this branch) then we can't trust anything that's returned from the cache.

That's a good point, while sandboxing should prevent them from accessing any network or the token, there's always a possibility that it will be broken.

I will move it into a seperate job and then export the created image as artefacts.

Reply:
I've moved it to a separate job, now it should be ok.